File: cpp/medium-solution/2523.cc
/*
2523. Closest Prime Numbers in Range
    Given two positive integers left and right, find the two integers num1 and num2 such that:
    left <= num1 < num2 <= right .
    Both num1 and num2 are prime numbers.
    num2 - num1 is the minimum amongst all other pairs satisfying the above conditions.
    Return the positive integer array ans = [num1, num2]. If there are multiple pairs satisfying these conditions, return the one with the smallest num1 value. If no such numbers exist, return [-1, -1].

    Example :
    Input: left = 10, right = 19
    Output: [11,13]
    Explanation: The prime numbers between 10 and 19 are 11, 13, 17, and 19.
    The closest gap between any pair is 2, which can be achieved by [11,13] or [17,19].
    Since 11 is smaller than 17, we return the first pair.
*/


#include <vector>
#include <climits>
using namespace std;
class Solution {
    public:
        vector<int> closestPrimes(int left, int right) {
            if (right < 2)
                return {-1, -1};
            vector<bool> isPrime(right + 1, true);
            isPrime[0] = isPrime[1] = false;
            for (int i = 2; i * i <= right; i++){
                if (isPrime[i]){
                    for (int j = i * i; j <= right; j+= i){
                        isPrime[j] = false;
                    }
                }
            }
            vector<int> primes;
            for (int i = left; i <= right; i++){
                if (isPrime[i])
                    primes.push_back(i);
            }
            if (primes.size() < 2)
                return {-1, -1};
            int minDiff = INT_MAX;
            vector<int> res = {-1, -1};
            for (size_t i = 1; i < primes.size(); i++){
                int diff = primes[i] - primes[i - 1];
                if (diff < minDiff){
                    minDiff = diff;
                    res = {primes[i - 1], primes[i]};
                }
            }
            return res;
        }
    };

File: python3/medium-solution/2523.py
"""
2523. Closest Prime Numbers in Range
    Given two positive integers left and right, find the two integers num1 and num2 such that:
    left <= num1 < num2 <= right .
    Both num1 and num2 are prime numbers.
    num2 - num1 is the minimum amongst all other pairs satisfying the above conditions.
    Return the positive integer array ans = [num1, num2]. If there are multiple pairs satisfying these conditions, return the one with the smallest num1 value. If no such numbers exist, return [-1, -1].

    Example :
    Input: left = 10, right = 19
    Output: [11,13]
    Explanation: The prime numbers between 10 and 19 are 11, 13, 17, and 19.
    The closest gap between any pair is 2, which can be achieved by [11,13] or [17,19].
    Since 11 is smaller than 17, we return the first pair.
"""

from typing import List
from math import sqrt

class Solution:
    def closestPrimes(self, left: int, right: int) -> List[int]:
        def isPrime(n : int) -> bool:
            if n <= 1:
                return False
            if n == 2:
                return True
            
            for i in range(2, int(sqrt(n)) + 1):
                if n % i == 0:
                    return False
            return True
        res = []
        prime = None

        for i in range(left, right + 1):
            if not isPrime(i):
                continue
            if prime is None:
                prime = i
                continue
            diff = i - prime
            if diff <= 2:
                return [prime, i]
            if not res or diff < res[1] - res[0]:
                res = [prime, i]
            prime = i
        return res or [-1, -1]